### PR TITLE
Fix all bugs

### LINKED ISSUE #65: Unhelpful error message when trying to write a ``TensorBlock`` to file
When trying to write an ``TensorBlock`` object to file using the ``equistore.io.save()`` function, the following error message is given:
```py
ArgumentError: argument 2: <class 'TypeError'>: expected LP_eqs_tensormap_t instance instead of LP_eqs_block_t
```

**Example**
```py
import numpy as np

from equistore import io, TensorBlock

# Create an example TensorBlock
block = TensorBlock(values=np.random.random((2, 2)),
                    samples=Labels(names=('structure','center'), values=np.array([[0, 1], [0, 2]])),
                    components=[],
                    properties=Labels(names=('n',), values=np.array([(0,), (1,),])),
                   )

# Attempt to save to file
io.save('tensorblock.npz', block)
>>> ArgumentError: argument 2: <class 'TypeError'>: expected LP_eqs_tensormap_t instance instead of LP_eqs_block_t
```

**Desired Behaviour**

The error message could be more helpful by explaining more explicitly / less cryptically that only ``TensorMap`` objects, and not ``TensorBlock`` objects, can be saved.

### LINKED ISSUE #174: Accessing blocks in empty tensor map
Hey! I think this is a bug. I am able to create an empty tensor map, but then I am not allowed to access the blocks as the code raises a ValueError `ValueError: Couldn't find any block matching the selection {}`

Snippet to reproduce the bug

```py
empty_tensor = TensorMap(keys=Labels.empty("key"), blocks=[])
empty_tensor.blocks()
```

### LINKED ISSUE #218: Make gradient blocks with samples that are not contained in the values invalid
This code works, even though sample 1 (referenced in the gradient block) is not in the values of the tensor block.
```python
block_1 = TensorBlock(
    values=np.full((1, 1), 1.0),
    samples=Labels(["samples"], np.array([[0]], dtype=np.int32)),
    components=[],
    properties=Labels(["properties"], np.array([[0]], dtype=np.int32)),
)
block_1.add_gradient(
    "parameter",
    samples=Labels(
        ["sample"], np.array([[1]], dtype=np.int32)
    ),
    data=np.full((1, 1), 11.0),
    components=[],
)
```

Code that needs to be modified 
https://github.com/lab-cosmo/equistore/blob/b8a65fa205ffc0542b0f34a58ace607e883a3271/equistore-core/src/blocks.rs#L304-L309

### PR DESCRIPTION
**Pull request by Luthaf**
Fix #65
Fix #174 
Fix #218 

----
:books: Documentation preview :books:: https://equistore--262.org.readthedocs.build/en/262/

### CODE REVIEW

File: python/equistore-core/equistore/core/labels.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ class Labels(np.ndarray):
             ...
     """
 
-    def __new__(cls, names: Union[List[str], str], values: np.ndarray, **kwargs):
+    def __new__(cls, names: Union[Sequence[str], str], values: np.ndarray, **kwargs):
```
Comment (jwa7) [inline python/equistore-core/equistore/core/labels.py:60]: Why the change from List to Sequence here?

File: python/equistore-core/tests/tensor.py [hunk 2]
```diff
@@ -361,3 +363,35 @@ def test_components_to_properties(tensor):
...
+    message = "invalid parameter: 'not_a_key' is not part of the keys for this tensor"
+    with pytest.raises(EquistoreError, match=message):
+        empty_tensor.blocks_matching(not_a_key=3)
+
+    message = (
+        "invalid parameter: block index out of bounds: we have "
+        "0 blocks but the index is 3"
+    )
+    with pytest.raises(EquistoreError, match=message):
+        empty_tensor.block(3)
```
Comment (jwa7) [inline python/equistore-core/tests/tensor.py:390]: Would be good to have a test on `blocks()` that also raises an error

Comment (Luthaf) [reply python/equistore-core/tests/tensor.py:409]: done!
